untyped-lambda-calculus/article.md

```diff
@@ -0,0 +1,549 @@
+The sub-expression does not contain any other redex, so is the innermost.
+
+Reducing the leftmost outermost redex first is called the _normal order strategy_, and reducing the leftmost innermost first is called the _applicative order strategy_.
+
+In other words, the applicative order strategy first reduces all of the arguments (from left to right) and then applies the abstraction, while normal order does the inverse.
+
+Expressions that are not reducible using the chosen evaluation strategy are said to be in the _normal form_.
+
+If a given term has a normal form under both of these strategies, it is the same for both.
+But the applicative order strategy is strictly weaker than the normal order strategy: some terms that normalize under normal order don't normalize under applicative order.
+
+#### Reducing abstraction bodies or not
+
+The two strategies above are free to go inside abstractions to find redexes.
+For this reason, we call them _strong_.
+
+But there are also strategies that can't go inside abstraction bodies, which we call _weak_.
+When we're talking about programming languages, we can't necessarily inspect function bodies, so the weak strategies are considered more practical there.
+
+The weak strategy corresponding to the normal order is called _call-by-name_, and the one corresponding to the applicative order is _call-by-value_.
+
+Call-by-name corresponds to lazy evaluation.
+In lazy evaluation, some arguments are potentially never evaluated at all, but some computations may be duplicated.
+Very few programming languages implement lazy evaluation this way (the only real example that comes to mind is Algol 60).
+
+A more practical variation of call-by-name is _call-by-need_, which is, at least in theory, used in Haskell.
+Call-by-need behaves exactly like call-by-name, except that reduction steps that would duplicate computations, don't.
+It does that by sharing the computation corresponding to an argument everywhere it appears, so any shared computation is performed at most once.
+
+Call-by-value, on the other hand, corresponds to eager evaluation.
+In eager evaluation, arguments are always evaluated before the function call.
+Most programming languages follow this strategy by default.
+
+To sum up, here's a table with the strategies mentioned and their characteristics:
+
+| Name                   | Outermost- or innermost- redex first? | Strong or weak? |
+|------------------------|---------------------------------------|-----------------|
+| Normal order           | Outermost                             | Strong          |
+| Applicative order      | Innermost                             | Strong          |
+| Call-by-name (lazy)    | Outermost                             | Weak            |
+| Call-by-value (eager)  | Innermost                             | Weak            |
+
+### Eta-reduction
+
+Speaking of reductions, let's also briefly mention _eta_-reduction ($\eta$-reduction).
```

Comment:
@lierdakil I put this is a specific subsection. It looks like it makes sense now. You might want to change the text a little, I obviously don't know what I am talking about or why it is relevant to Haskell users. 😂